### PR TITLE
Add response body in failed HTTP Probe message

### LINKED ISSUE #103592: Add response body in failed HTTP Probe message
#### What would you like to be added:

It would be helpful to see a descriptive message with response body if HTTP probe failed.

Currently kubernetes returns body as message only for succeeded HTTP probes.
https://github.com/kubernetes/kubernetes/blob/d80e3d104c1f758df644dad38d007f692704baad/pkg/probe/http/http.go#L128-L138

#### Why is this needed:

If HTTP probe failed, it is useful and convenient to find out the reason from its message. At the moment it is unclear what is wrong with the container from HTTP probe message.

### PR DESCRIPTION
**Pull request by chenyw1990**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:

> It would be helpful to see a descriptive message with response body if HTTP probe failed.
https://github.com/kubernetes/kubernetes/issues/103592#issue-940161027

#### Which issue(s) this PR fixes:
Fixes #103592

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/kubelet/prober/prober.go [hunk 1]
```diff
@@ -110,10 +110,10 @@ func (pb *prober) probe(probeType probeType, pod *v1.Pod, status v1.PodStatus, c
 	if err != nil || (result != probe.Success && result != probe.Warning) {
 		// Probe failed in one way or another.
 		if err != nil {
-			klog.V(1).ErrorS(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
+			klog.ErrorS(err, "Probe errored", "probeType", probeType, "pod", klog.KObj(pod), "podUID", pod.UID, "containerName", container.Name)
```
Comment (wangyysde) [inline pkg/kubelet/prober/prober.go:113]: Could you tell me why we should modify this to “Add response body in failed HTTP Probe message”？

Comment (chenyw1990) [reply pkg/kubelet/prober/prober.go]: When checking whether output is print to in logs, I found that the two codes can be optimized. V(1) is unnecessary. Therefore, I changed the code accordingly.

I've reverted this change and re-introduced a new Pull Request to fix the code of klog.V(1).

File: pkg/probe/http/http.go [hunk 2]
```diff
@@ -135,7 +135,7 @@ func DoHTTPProbe(url *url.URL, headers http.Header, client GetHTTPInterface) (pr
 		return probe.Success, body, nil
 	}
 	klog.V(4).Infof("Probe failed for %s with request headers %v, response body: %v", url.String(), headers, body)
-	return probe.Failure, fmt.Sprintf("HTTP probe failed with statuscode: %d", res.StatusCode), nil
+	return probe.Failure, fmt.Sprintf("HTTP probe failed with statuscode: %d, response body: %v", res.StatusCode, body), nil
```
Comment (endocrimes) [inline pkg/probe/http/http.go:138]: We might want to limit how much of a body we're willing to return here? - A lot of web frameworks will return a bunch of garbage that we probably don't want to shovel around (and wouldn't be useful anyway). - Not sure what the reasonable limit would be though tbh

**Comment by aojea** [comment]
These will make this security issue even worse https://github.com/kubernetes/kubernetes/issues/99425#issue-815926103
/assign @tallclair

File: pkg/probe/http/http.go (hunk 2 above)
Comment (ehashman) [reply pkg/probe/http/http.go:138]: Yeah I am not sure we want to sprintf the entire response body without truncating. Doesn't this also have the potential to leak applications' internal state?

/hold

**Review by ehashman — CHANGES_REQUESTED** [review]

**Comment by tallclair** [comment]
Closing for potential information exposure.

**Comment by neiljerram** [comment]
Cool, just checking I understand correctly:
- The security issue is with HTTP probes in general, because they can be configured to point anywhere, so basically it's a way of getting kubelet to do an arbitrary HTTP query for you, and reveal the result.
- In comparison, an Exec probe (which does already include output in the failure case) is considered safer because it is restricted to exec'ing something that the pod image has provided.

@tallclair @aojea @endocrimes Is that understanding correct?

File: pkg/probe/http/http.go (hunk 2 above)
Comment (neiljerram) [reply pkg/probe/http/http.go:138]: This is moot now, but, for the record, the `body` has already been truncated to `maxRespBodyLength`, a few lines above.

**Comment by neiljerram** [comment]
Oops, have commented on the wrong (very similar looking!) PR - will repost at #108918 .

**Pull request closed without merging**